src/LinuxTracing/LibunwindstackMultipleOfflineAndProcessMemoryTest.cpp

```diff
@@ -0,0 +1,182 @@
+
+  std::array<char, 3> destination{};
+  destination.fill(0x11);
+  size_t read_count = sut->Read(kStartAddress - 1, destination.data(), 3);
+
+  ASSERT_EQ(read_count, 0);
+  EXPECT_EQ(destination[0], 0x11);
+  EXPECT_EQ(destination[1], 0x11);
+  EXPECT_EQ(destination[2], 0x11);
+}
+
+TEST(LibunwindstackMultipleOfflineAndProcessMemory,
+     RequestingToReadUnknownMemoryWithoutProcessReturnsZero) {
+  constexpr uint64_t kStartAddress = 0xADD8E55;
+  std::vector<char> bytes{0x01, 0x10, 0x20, 0x30, 0x40};
+  StackSliceView stack_slice(kStartAddress, bytes.size(), bytes.data());
+
+  std::shared_ptr<unwindstack::Memory> sut =
+      LibunwindstackMultipleOfflineAndProcessMemory::CreateWithoutProcessMemory({stack_slice});
+
+  std::array<char, 3> destination{};
+  destination.fill(0x11);
+  size_t read_count = sut->Read(0xFE, destination.data(), 3);
+
+  ASSERT_EQ(read_count, 0);
+  EXPECT_EQ(destination[0], 0x11);
+  EXPECT_EQ(destination[1], 0x11);
+  EXPECT_EQ(destination[2], 0x11);
+}
+
+TEST(LibunwindstackMultipleOfflineAndProcessMemory,
+     ReadFromCompleteMemoryEvenIfPartiallyIntersectsWithOtherStackSlice) {
+  constexpr uint64_t kStartAddress1 = 0xADD8E55;
+  std::vector<char> bytes1{0x01, 0x10, 0x20, 0x30, 0x40};
+  StackSliceView stack_slice1(kStartAddress1, bytes1.size(), bytes1.data());
+
+  constexpr uint64_t kStartAddress2 = kStartAddress1 + 2;
+  std::vector<char> bytes2{0x20, 0x30, 0x40, 0x50, 0x60, 0x70, 0x77};
+  StackSliceView stack_slice2(kStartAddress2, bytes2.size(), bytes2.data());
+
+  std::shared_ptr<unwindstack::Memory> sut =
+      LibunwindstackMultipleOfflineAndProcessMemory::CreateWithoutProcessMemory(
+          {stack_slice1, stack_slice2});
+
+  std::array<char, 3> destination{};
+  size_t read_count = sut->Read(kStartAddress2, destination.data(), 3);
+
+  ASSERT_EQ(read_count, 3);
+  EXPECT_EQ(destination[0], 0x20);
+  EXPECT_EQ(destination[1], 0x30);
+  EXPECT_EQ(destination[2], 0x40);
+}
+
+TEST(LibunwindstackMultipleOfflineAndProcessMemory, ReadFromProcess) {
+  constexpr uint64_t kStartAddress1 = 0xADD8E55;
+  std::vector<char> bytes1{0x01, 0x10, 0x20, 0x30, 0x40};
+  StackSliceView stack_slice1(kStartAddress1, bytes1.size(), bytes1.data());
+
+  constexpr uint64_t kStartAddress2 = 0xABCDEF;
+  std::vector<char> bytes2{0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17};
+  StackSliceView stack_slice2(kStartAddress2, bytes2.size(), bytes2.data());
+
+  std::vector<StackSliceView> stack_slices{stack_slice1, stack_slice2};
+  std::vector<LibunwindstackOfflineMemory> stack_memory_slices{};
+  stack_memory_slices.reserve(stack_slices.size());
+  for (const StackSliceView& stack_slice_view : stack_slices) {
+    stack_memory_slices.emplace_back(stack_slice_view);
+  }
+
+  std::shared_ptr<MockProcessMemory> process_memory_mock = std::make_shared<MockProcessMemory>();
+  LibunwindstackMultipleOfflineAndProcessMemory sut{process_memory_mock,
```

Comment:
So, I had this idea, I'll leave it here and you decide what to do with it.

The constructor of `LibunwindstackMultipleOfflineAndProcessMemory`, and the `LibunwindstackOfflineMemory` class, could both be private in `LibunwindstackMultipleOfflineAndProcessMemory`... except for this test.

A way around it would be to use `CreateWithProcessMemory`, pass the pid of the current process and the `stack_slices`, and then `Read` at the address that corresponds to a pointer to a local variable (array). This would also get rid of `MockProcessMemory`.

I think it would clean up what we expose from `LibunwindstackMultipleOfflineAndProcessMemory.h`.

Reply:
I thought about that before and originally decided that this would be too much of an integration test. But as I was kind of forced to do the integration test approach already for the offline memory, I think it is actually cleaner to also do it for the process. The test code becomes also much more readable.